### PR TITLE
Challenge1 | publish a letter to the SNS topic

### PR DESCRIPTION
**Pull request by CristinaHA93**
Screenshot 2024-12-10 100549

### CODE REVIEW

File: letter-sender/src/main/java/com/christmas/letter/sender/SampleController.java [hunk 1]
```diff
@@ -1,16 +1,30 @@
...
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
 import org.springframework.web.bind.annotation.RequestMapping;
 import org.springframework.web.bind.annotation.RestController;
 
 @RestController
-@RequestMapping("/sample")
+@RequestMapping("/api/letters")
+@RequiredArgsConstructor
 public class SampleController {
```
Comment (teodora-bobirneci) [inline letter-sender/src/main/java/com/christmas/letter/sender/SampleController.java]: Let's rename this to something more friendly, maybe LetterController.

File: letter-sender/src/main/java/com/christmas/letter/sender/exceptions/LetterControllerAdvice.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+import java.util.Date;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.context.request.WebRequest;
+import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
+
+@ControllerAdvice
+public class LetterControllerAdvice extends ResponseEntityExceptionHandler {
```
Comment (teodora-bobirneci) [inline letter-sender/src/main/java/com/christmas/letter/sender/exceptions/LetterControllerAdvice.java]: Let's give this component a better name, I wouldn't use Advice as that is only an implementation detail.

File: letter-sender/src/main/java/com/christmas/letter/sender/exceptions/LetterValidator.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+import com.christmas.letter.sender.model.Letter;
+import java.util.ArrayList;
+import java.util.List;
+import org.springframework.stereotype.Component;
+
+@Component
+public class LetterValidator {
+
+  public void validate(Letter letter){
```
Comment (teodora-bobirneci) [inline letter-sender/src/main/java/com/christmas/letter/sender/exceptions/LetterValidator.java]: This isn't used, however, I suggest 2 possible approaches here - reference for the future, no need to waste time on it unless you want to. 
1. I would create an annotation, for example @ValidCoordinate or something similar, and I would move this code in a validator linked to that annotation. If you want to go this path and can't find an example, let me know.
2. I would simplify how we store the coordinates - make it a string and store the address instead, or still make it a string and use the above recommendation.  

And if we go this path, should we allow negative values for example?

File: letter-sender/src/main/java/com/christmas/letter/sender/model/Letter.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package com.christmas.letter.sender.model;
+
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class Letter {
+
+  private String email;
```
Comment (teodora-bobirneci) [inline letter-sender/src/main/java/com/christmas/letter/sender/model/Letter.java:16]: Let's use annotations here to make sure the user cannot provide an empty email, or name, etc

File: letter-sender/src/main/java/com/christmas/letter/sender/service/SnsService.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+import java.util.regex.Pattern;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+import software.amazon.awssdk.services.sns.SnsClient;
+import software.amazon.awssdk.services.sns.model.PublishRequest;
+
+@Service
+@RequiredArgsConstructor
+public class SnsService {
```
Comment (teodora-bobirneci) [inline letter-sender/src/main/java/com/christmas/letter/sender/service/SnsService.java]: The fact that this service uses SNS is an implementation detail. Let's find a more suitable name for it, describing what it does, not how it does it.

File: letter-sender/src/main/java/com/christmas/letter/sender/service/SnsService.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+  @NonNull
```
Comment (teodora-bobirneci) [inline letter-sender/src/main/java/com/christmas/letter/sender/service/SnsService.java]: This is just a preference, but there's no need to keep adding @NonNull wherever we inject something. If such code reaches production, it's clear no one tested even the slightest functionality, so the nonnull part is the smallest concern.

File: letter-sender/src/main/java/com/christmas/letter/sender/service/SnsService.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  private final SnsClient snsClient;
+  private final String snsTopicArn = "arn:aws:sns:us-east-1:000000000000:ChristmasLetterSns";
+  private static final Pattern EMAIL_PATTERN = Pattern.compile(
+      "^[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+\\.[a-zA-Z]{2,6}$");
+
+  public String publishLetter(Letter letter) {
+
+      if (letter.getEmail() == null || !EMAIL_PATTERN.matcher(letter.getEmail()).matches()) {
```
Comment (teodora-bobirneci) [inline letter-sender/src/main/java/com/christmas/letter/sender/service/SnsService.java]: Let's move the validations outside the service layer. You can specify the regex along with the nonnull validation in the model itself, via annotations.
See https://www.javaguides.net/2021/04/spring-boot-dto-validation-example.html

File: letter-sender/src/main/java/com/christmas/letter/sender/service/SnsService.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        throw new InvalidDataException("Invalid email address provided: " + letter.getEmail());
+      }
+
+      String message = buildLetterMessage(letter);
```
Comment (teodora-bobirneci) [inline letter-sender/src/main/java/com/christmas/letter/sender/service/SnsService.java]: I would suggest publishing directly as json, no need to process it.

File: letter-sender/src/main/resources/localstack-setup.sh [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+until awslocal sns list-topics &> /dev/null; do
+echo "Waiting for LocalStack..."
+sleep 5
+done
+
+echo "Adding example item to the DynamoDB table..."
+awslocal dynamodb put-item \
+--table-name SantaChristmasLetters \
+--item '{
+"Email": {"S": "cristina.stoenescu93@gmail.com"},
```
Comment (teodora-bobirneci) [inline letter-sender/src/main/resources/localstack-setup.sh]: If this is your real email, I wouldn't put it in a public github repository.

File: letter-sender/src/test/java/com/christmas/letter/sender/service/SnsServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+import software.amazon.awssdk.services.sns.model.PublishResponse;
+
+import static org.hamcrest.MatcherAssert.assertThat;;
+
+import static org.hamcrest.Matchers.notNullValue;
+import static org.junit.Assert.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.Mockito.*;
+
+@SpringBootTest
```
Comment (teodora-bobirneci) [inline letter-sender/src/test/java/com/christmas/letter/sender/service/SnsServiceTest.java]: @SpringBootTest is used for integration testing, however, this is clearly a unit test - you are mocking the dependencies and testing the service in isolation. So, you can you @ExtendWith(MockitoExtension.class

File: letter-sender/src/test/java/com/christmas/letter/sender/service/SnsServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+  void When_CreateLetter_Expect_LetterPublished() {
+
+    PublishResponse mockResponse = PublishResponse.builder()
+        .build();
+
+    when(snsClient.publish(any(PublishRequest.class))).thenReturn(mockResponse);
+
+    String returned = snsService.publishLetter(createLetter());
+
+    assertThat(returned, notNullValue());
```
Comment (teodora-bobirneci) [inline letter-sender/src/test/java/com/christmas/letter/sender/service/SnsServiceTest.java]: We should also validate that the letter has all the elements we're interested in, not just that it is not null.

File: letter-sender/src/test/java/com/christmas/letter/sender/service/SnsServiceTest.java [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+
+  }
+
+  @Test
+  void When_CreateLetter_Expect_failure() {
```
Comment (teodora-bobirneci) [inline letter-sender/src/test/java/com/christmas/letter/sender/service/SnsServiceTest.java]: Let's add tests that validate that no other fields can have invalid data, for example negative coordinates.

**Review by teodora-bobirneci — COMMENTED** [review]
Let's also add some integration tests, I would suggest using the MockMVC approach, see an example here:
https://spring.io/guides/gs/testing-web

File: letter-sender/src/main/java/com/christmas/letter/sender/SampleController.java (hunk 1 above)
Comment (CristinaHA93) [reply letter-sender/src/main/java/com/christmas/letter/sender/SampleController.java]: changed to LetterController

File: letter-sender/src/main/java/com/christmas/letter/sender/exceptions/LetterControllerAdvice.java (hunk 2 above)
Comment (CristinaHA93) [reply letter-sender/src/main/java/com/christmas/letter/sender/exceptions/LetterControllerAdvice.java]: name updated to Advice

File: letter-sender/src/main/java/com/christmas/letter/sender/model/Letter.java (hunk 4 above)
Comment (CristinaHA93) [reply letter-sender/src/main/java/com/christmas/letter/sender/model/Letter.java:16]: solved with:
  @NotNull(message="Email required")
  @Email(regexp = "[a-z0-9._%+-]+@[a-z0-9.-]+\\.[a-z]{2,3}")
  private String email;

File: letter-sender/src/main/java/com/christmas/letter/sender/service/SnsService.java (hunk 5 above)
Comment (CristinaHA93) [reply letter-sender/src/main/java/com/christmas/letter/sender/service/SnsService.java]: changed to LetterService

File: letter-sender/src/main/resources/localstack-setup.sh (hunk 9 above)
Comment (CristinaHA93) [reply letter-sender/src/main/resources/localstack-setup.sh]: update it

File: letter-sender/src/main/java/com/christmas/letter/sender/service/SnsService.java (hunk 8 above)
Comment (CristinaHA93) [reply letter-sender/src/main/java/com/christmas/letter/sender/service/SnsService.java]: update using ObjectMapper
String message = objectMapper.writeValueAsString(letter);

File: letter-sender/src/test/java/com/christmas/letter/sender/service/SnsServiceTest.java (hunk 10 above)
Comment (CristinaHA93) [reply letter-sender/src/test/java/com/christmas/letter/sender/service/SnsServiceTest.java]: Removed @SpringBootTest use @ExtendWith(MockitoExtension.class)

File: letter-sender/src/test/java/com/christmas/letter/sender/service/SnsServiceTest.java (hunk 11 above)
Comment (CristinaHA93) [reply letter-sender/src/test/java/com/christmas/letter/sender/service/SnsServiceTest.java]: I create another test where I check if the letter is properly set

File: letter-sender/src/main/java/com/christmas/letter/sender/exceptions/LetterValidator.java (hunk 3 above)
Comment (CristinaHA93) [reply letter-sender/src/main/java/com/christmas/letter/sender/exceptions/LetterValidator.java]: I called this validator method in service method.
on mape are some locations where the coordinates are negative, example: 21.3069°, -157.8583°

File: letter-sender/src/test/java/com/christmas/letter/sender/service/SnsServiceTest.java (hunk 12 above)
Comment (CristinaHA93) [reply letter-sender/src/test/java/com/christmas/letter/sender/service/SnsServiceTest.java]: add 2 test where I check if the letter have invalid email and invalide coordinates

File: letter-sender/src/main/java/com/christmas/letter/sender/LetterController.java [hunk 13]
```diff
@@ -0,0 +1,30 @@
+package com.christmas.letter.sender;
```
Comment (teodora-bobirneci) [inline letter-sender/src/main/java/com/christmas/letter/sender/LetterController.java]: Let's move this class to a dedicated package. You can use controller or rest or whatever name you see fit.

File: letter-sender/src/main/java/com/christmas/letter/sender/exceptions/LetterControllerAdvice.java (hunk 2 above)
Comment (teodora-bobirneci) [reply letter-sender/src/main/java/com/christmas/letter/sender/exceptions/LetterControllerAdvice.java]: I was saying not to use advice in the class name. This class is a central place where we handle common exceptions. A generally used name is GenericExceptionHandler or GenericErrorHandler.

File: letter-sender/src/main/java/com/christmas/letter/sender/exceptions/LetterValidator.java (hunk 3 above)
Comment (teodora-bobirneci) [reply letter-sender/src/main/java/com/christmas/letter/sender/exceptions/LetterValidator.java]: This would be a cleaner way to do the validation: https://www.baeldung.com/spring-mvc-custom-validator

File: letter-sender/src/main/java/com/christmas/letter/sender/service/SnsService.java (hunk 7 above)
Comment (teodora-bobirneci) [reply letter-sender/src/main/java/com/christmas/letter/sender/service/SnsService.java]: We already have these validations on the Letter object, do we need to do the same check here as well?  
@NotNull(message="Email required")
  @Email(regexp = "[a-z0-9._%+-]+@[a-z0-9.-]+\\.[a-z]{2,3}")

File: letter-sender/src/main/java/com/christmas/letter/sender/service/LetterService.java [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+@RequiredArgsConstructor
+public class LetterService {
+
+  @NonNull
+  private final SnsClient snsClient;
+  @NonNull
+  private final LetterValidator letterValidator;
+  private final ObjectMapper objectMapper;
+
+  private final String snsTopicArn = "arn:aws:sns:us-east-1:000000000000:ChristmasLetterSns";
```
Comment (teodora-bobirneci) [inline letter-sender/src/main/java/com/christmas/letter/sender/service/LetterService.java]: For this to be a constant, it should also be static, and let's update the name to follow the constant naming conventions, capital case only and _ as separator.

File: letter-sender/src/main/java/com/christmas/letter/sender/service/LetterService.java [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+
+  public String publishLetter(Letter letter) {
+
+    try{
+
+      this.letterValidator.validate(letter);
+
+      String message = objectMapper.writeValueAsString(letter);
+
+      PublishRequest publishRequest = PublishRequest.builder()
```
Comment (teodora-bobirneci) [inline letter-sender/src/main/java/com/christmas/letter/sender/service/LetterService.java]: It is ok as is, but if you are curious, there is also an alternative, you can have a look at snsTemplate.convertAndSend method, from io.awspring.cloud.

File: letter-sender/src/test/java/com/christmas/letter/sender/service/SnsServiceTest.java [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+import static org.hamcrest.MatcherAssert.assertThat;;
+
+import static org.hamcrest.Matchers.notNullValue;
+import static org.junit.Assert.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.Mockito.*;
+
+@ExtendWith(MockitoExtension.class)
+class SnsServiceTest {
```
Comment (teodora-bobirneci) [inline letter-sender/src/test/java/com/christmas/letter/sender/service/SnsServiceTest.java]: The service has a new name, let's also rename this test.

**Review by teodora-bobirneci — COMMENTED** [review]
The test you have right now is a unit test. Let's also add some integration tests. This is a good starting point: https://github.com/IrinaButaru/christmas-letter-handling/blob/957f051c2b79e25968e93d7b710e424dc5470b89/letter-sender/src/test/java/com/christmas/letter/sender/controller/LetterSenderControllerIntegrationTest.java

File: letter-sender/src/main/java/com/christmas/letter/sender/LetterController.java (hunk 13 above)
Comment (CristinaHA93) [reply letter-sender/src/main/java/com/christmas/letter/sender/LetterController.java]: refactor the controller class in controller package

File: letter-sender/src/test/java/com/christmas/letter/sender/service/SnsServiceTest.java (hunk 16 above)
Comment (CristinaHA93) [reply letter-sender/src/test/java/com/christmas/letter/sender/service/SnsServiceTest.java]: updated to LetterServiceTest

File: letter-sender/src/main/java/com/christmas/letter/sender/configuration/AwsConfig.java [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+
+@Configuration
+public class AwsConfig {
+
+  @Bean
+  public SnsClient configSns(){
+    return SnsClient.builder()
+        .region(Region.US_EAST_1)
+        .endpointOverride(URI.create("http://localhost:4566"))
+        .build();
```
Comment (tudorcodrea) [inline letter-sender/src/main/java/com/christmas/letter/sender/configuration/AwsConfig.java]: these properties usually change over time, they can be loaded from application.properties files

**Review by tudorcodrea — COMMENTED** [review]
looks good, I would load all external (resource dependent) properties from properties files

File: letter-sender/src/main/java/com/christmas/letter/sender/service/LetterService.java (hunk 14 above)
Comment (CristinaHA93) [reply letter-sender/src/main/java/com/christmas/letter/sender/service/LetterService.java]: updated to private static final String Sns_Topic_Arn

File: letter-sender/src/main/java/com/christmas/letter/sender/configuration/AwsConfig.java (hunk 17 above)
Comment (CristinaHA93) [reply letter-sender/src/main/java/com/christmas/letter/sender/configuration/AwsConfig.java]: I moved the endpoint in application.properties

File: letter-sender/src/main/java/com/christmas/letter/sender/service/LetterService.java (hunk 15 above)
Comment (CristinaHA93) [reply letter-sender/src/main/java/com/christmas/letter/sender/service/LetterService.java]: refactor the service using snsTemplate.convertAndSend

File: letter-sender/src/main/java/com/christmas/letter/sender/exceptions/LetterControllerAdvice.java (hunk 2 above)
Comment (teodora-bobirneci) [reply letter-sender/src/main/java/com/christmas/letter/sender/exceptions/LetterControllerAdvice.java]: This is still pending.

File: letter-sender/src/main/java/com/christmas/letter/sender/exceptions/LetterValidator.java [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+import org.springframework.stereotype.Component;
+
+@Component
+public class LetterValidator{
+  private static final Pattern EMAIL_PATTERN = Pattern.compile(
+      "^[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+\\.[a-zA-Z]{2,6}$");
+
+  public void validate(Letter letter) {
+
+    if (letter.getEmail() == null || !EMAIL_PATTERN.matcher(letter.getEmail()).matches()) {
```
Comment (teodora-bobirneci) [inline letter-sender/src/main/java/com/christmas/letter/sender/exceptions/LetterValidator.java]: isn't the email format already being validated with the @Email annotation from the Letter class?

File: letter-sender/src/test/java/com/christmas/letter/sender/controller/LetterControllerTest.java [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.http.MediaType;
+import org.springframework.test.web.servlet.MockMvc;
+
+@WebMvcTest(LetterController.class)
+class LetterControllerTest {
+
+  @SuppressWarnings("SpringJavaInjectionPointsAutowiringInspection")
```
Comment (teodora-bobirneci) [inline letter-sender/src/test/java/com/christmas/letter/sender/controller/LetterControllerTest.java]: Do we need this suppress annotation?

File: letter-sender/src/test/java/com/christmas/letter/sender/controller/LetterControllerTest.java [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+        .andExpect(status().isCreated());
+
+  }
+
+ @Test
+  void When_CreateLetter_Expect_BadRequest() throws Exception {
+
+    String invalidTestMessage = objectMapper.writeValueAsString(createInvalidLetter());
+
+    mockMvc.perform(post(PATH)
```
Comment (teodora-bobirneci) [inline letter-sender/src/test/java/com/christmas/letter/sender/controller/LetterControllerTest.java:47]: Let's also verify that the message returned by the api is helping the user correct the payload, so let's verify that it actually says "email is required" etc etc

File: letter-sender/src/test/java/com/christmas/letter/sender/service/LetterServiceTest.java [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+    letter.setEmail("invalid-email");
+    letter.setName("Cristina");
+    letter.setWishes("Dear Santa, I wish to visit you this year.");
+
+    Location location = new Location();
+    location.setLatitude(45.760696);
+    location.setLongitude(-74.0060);
+    letter.setLocation(location);
+
+    Set<ConstraintViolation<Letter>> violations = validator.validate(letter);
```
Comment (teodora-bobirneci) [inline letter-sender/src/test/java/com/christmas/letter/sender/service/LetterServiceTest.java]: This test verifies that the validator implementationfrom Jakarta knows how to do its job. Our goal is to make sure our own components do what they are supposed to. So we should be testing those.

File: letter-sender/src/test/java/com/christmas/letter/sender/controller/LetterControllerTest.java (hunk 19 above)
Comment (CristinaHA93) [reply letter-sender/src/test/java/com/christmas/letter/sender/controller/LetterControllerTest.java]: yes because do not recognized the @Autowired

File: letter-sender/src/main/java/com/christmas/letter/sender/exceptions/LetterValidator.java (hunk 18 above)
Comment (CristinaHA93) [reply letter-sender/src/main/java/com/christmas/letter/sender/exceptions/LetterValidator.java]: I removed the LetterValidator.class

File: letter-sender/src/test/java/com/christmas/letter/sender/controller/LetterControllerTest.java (hunk 19 above)
Comment (CristinaHA93) [reply letter-sender/src/test/java/com/christmas/letter/sender/controller/LetterControllerTest.java]: I remove it but sonar compain about @Autowired maybe is because maven version

File: letter-sender/src/test/java/com/christmas/letter/sender/service/LetterServiceTest.java (hunk 21 above)
Comment (CristinaHA93) [reply letter-sender/src/test/java/com/christmas/letter/sender/service/LetterServiceTest.java]: remove it and check the validation  in LetterControllerTest.java

File: letter-sender/src/test/java/com/christmas/letter/sender/controller/LetterControllerTest.java (hunk 20 above)
Comment (CristinaHA93) [reply letter-sender/src/test/java/com/christmas/letter/sender/controller/LetterControllerTest.java:47]: check .andExpect(jsonPath("$.message").value("The validation has failed!"));

**Review by tudorcodrea — APPROVED** [review]

**Review by teodora-bobirneci — APPROVED** [review]
Looks great ❤️
